src/sync/app.cpp

```diff
@@ -16,67 +16,71 @@
+, m_app_route(m_base_route + default_app_path + "/" + config.app_id)
+, m_auth_route(m_app_route + default_auth_path)
+, m_request_timeout_ms(config.default_request_timeout_ms.value_or(default_timeout_ms))
+{
+    REALM_ASSERT(m_config.transport_generator);
+}
 
-static std::unique_ptr<error::AppError> handle_error(const Response response) {
+static Optional<AppError> handle_error(const Response& response) {
     if (response.body.empty()) {
         return {};
     }
 
-    if (response.headers.find("Content-Type") != response.headers.end()) {
-        if (response.headers.at("Content-Type") == "application/json") {
-            auto body = nlohmann::json::parse(response.body);
-            return std::make_unique<error::AppError>(app::error::ServiceError(body["errorCode"].get<std::string>(),
-                                  body["error"].get<std::string>()));
-        }
-    }
-
-    if (response.custom_status_code != 0) {
-        return std::make_unique<error::AppError>(response.body,
-                                                 response.custom_status_code,
-                                                 error::AppError::Type::Custom);
+    if (auto ct = response.headers.find("Content-Type"); ct != response.headers.end() && ct->second == "application/json") {
+        auto body = nlohmann::json::parse(response.body);
+        return AppError(make_error_code(service_error_code_from_string(body["errorCode"].get<std::string>())),
+                              body["error"].get<std::string>());
     }
 
     return {};
 }
 
-inline bool response_code_is_fatal(Response response) {
+inline bool response_code_is_fatal(Response const& response) {
     // FIXME: our tests currently only generate codes 0 and 200,
     // but we need more robust error handling here; eg. should a 300
     // redirect really be considered fatal or should we automatically redirect?
-    return response.http_status_code >= 300 || response.http_status_code < 200 || response.custom_status_code != 0;
+    return response.http_status_code >= 300
+        || (response.http_status_code < 200 && response.http_status_code != 0)
+        || response.custom_status_code != 0;
 }
 
-void App::login_with_credentials(const std::shared_ptr<AppCredentials> credentials,
-                                 std::function<void(std::shared_ptr<SyncUser>, std::unique_ptr<error::AppError>)> completion_block) {
+void App::login_with_credentials(const AppCredentials& credentials,
+                                 std::function<void(std::shared_ptr<SyncUser>, Optional<AppError>)> completion_block) {
     // construct the route
-    std::string route = util::format("%1/providers/%2/login", m_auth_route, provider_type_from_enum(credentials->m_provider));
+    std::string route = util::format("%1/providers/%2/login", m_auth_route, credentials.provider_as_string());
 
-    auto handler = [&](const Response response) {
+    auto handler = [&](const Response& response) {
```

Comment:
This needs to capture `completion_block` by value rather than by reference as `handler` may be called after we return from this function.